src/VisualStudio/Core/Def/ProjectSystem/Legacy/SolutionEventsBatchScopeCreator.cs

```diff
@@ -4,245 +4,224 @@
+                    _fullSolutionLoadScopes.RemoveAt(i);
+
+                    // Go back by one so we re-check the same index
+                    i--;
+                }
             }
-        }
 
-        EnsureUnsubscribedFromRunningDocumentTableEventsIfNoLongerNeeded();
+            EnsureUnsubscribedFromRunningDocumentTableEventsIfNoLongerNeeded();
+        }
     }
 
     private void EnsureSubscribedToSolutionEvents()
     {
-        _threadingContext.ThrowIfNotOnUIThread();
-
-        if (_isSubscribedToSolutionEvents)
+        lock (_gate)
         {
-            return;
-        }
+            if (_isSubscribedToSolutionEvents)
+            {
+                return;
+            }
 
-        var solution = (IVsSolution)_serviceProvider.GetService(typeof(SVsSolution));
+            var backgroundSolution = (IVsBackgroundSolution)_serviceProvider.GetService(typeof(SVsBackgroundSolution));
 
-        // We never unsubscribe from these, so we just throw out the cookie. We could consider unsubscribing if/when all our
-        // projects are unloaded, but it seems fairly unnecessary -- it'd only be useful if somebody closed one solution but then
-        // opened other solutions in entirely different languages from there.
-        if (ErrorHandler.Succeeded(solution.AdviseSolutionEvents(new SolutionEventsEventSink(this), out _)))
-        {
-            _isSubscribedToSolutionEvents = true;
-        }
+            // We never unsubscribe from these, so we just throw out the subscription. We could consider unsubscribing if/when all our
+            // projects are unloaded, but it seems fairly unnecessary -- it'd only be useful if somebody closed one solution but then
+            // opened other solutions in entirely different languages from there.
 
-        // It's possible that we're loading after the solution has already fully loaded, so see if we missed the event 
-        var shellMonitorSelection = (IVsMonitorSelection)_serviceProvider.GetService(typeof(SVsShellMonitorSelection));
+            // The IVsAsyncSolutionEventListener subscription also ensures we'll receive an after-solution-loaded event if the solution is already
+            // loaded, so we don't need to check if we've missed the event.
+            _ = backgroundSolution.SubscribeListener(new SolutionEventsEventListener(this));
 
-        if (ErrorHandler.Succeeded(shellMonitorSelection.GetCmdUIContextCookie(VSConstants.UICONTEXT.SolutionExistsAndFullyLoaded_guid, out var fullyLoadedContextCookie)))
-        {
-            if (ErrorHandler.Succeeded(shellMonitorSelection.IsCmdUIContextActive(fullyLoadedContextCookie, out var fActive)) && fActive != 0)
-            {
-                _solutionLoaded = true;
-            }
+            _isSubscribedToSolutionEvents = true;
         }
     }
 
     private void EnsureSubscribedToRunningDocumentTableEvents()
     {
-        _threadingContext.ThrowIfNotOnUIThread();
```

Comment:
When this code was written, the running document table had UI thread affinity. It no longer has that for clients that are subscribing/unsubscribing changes, so there's no reason to assert this. Similarly for the use of the new background solution service, so at this point we don't need any further assertions.